yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history.yarn
+
+import java.net.{ConnectException, URI}
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+
+import scala.collection.mutable.LinkedList
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.security.UserGroupInformation
+import org.apache.hadoop.service.{AbstractService, Service}
+import org.apache.hadoop.yarn.api.records.{ApplicationAttemptId, ApplicationId}
+import org.apache.hadoop.yarn.api.records.timeline.{TimelineDomain, TimelineEntity, TimelinePutResponse}
+import org.apache.hadoop.yarn.client.api.TimelineClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import org.apache.spark.deploy.history.yarn.YarnTimelineUtils._
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster.YarnExtensionService
+import org.apache.spark.util.{SystemClock, Clock, Utils}
+import org.apache.spark.{Logging, SparkContext}
+
+/**
+ * Implements a Hadoop service with the init/start logic replaced by that
+ * of the YarnService.
+ * <p>
+ * As [[AbstractService]] implements `close()`, routing
+ * to its `stop` method, calling `close()` is sufficient
+ * to stop the service instance.
+ * <p>
+ * However, when registered to receive spark events, the service will continue to
+ * receive them until the spark context is stopped. Events received when this service
+ * is in a `STOPPED` state will be discarded.
+ */
+private[spark] class YarnHistoryService  extends AbstractService("History Service")
```

Comment:
1. It makes it easier to aggregate up things, and it has a robust model for thread-safe subclassing through a lifecycle. Most of YARN is a composition of these; in fact the spark YARN client and AM could adopt it for better management of `AMRMClient`, `YarnClient`, `YarnExtensionServices` & the like

2. The hive-thriftserver code *almost* uses it. More specifically, it used a cut-and-paste of the Hadoop 2.04-alpha era and then abuses java reflection to bypass bits of the direct parent. Once Hive moves to Hadoop 2.2+ only then it can be moved back into the YARN model —though once you modify Hive 1.2.x to provide override points the spark-hivethriftserver module doesn't need the `ReflectedCompositeService` hack any more.